components/permissions/google_sign_in/google_sign_in_permission_util.cc

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/permissions/google_sign_in/google_sign_in_permission_util.h"
+
+#include <utility>
+#include <vector>
+
+#include "base/callback_helpers.h"
+#include "base/functional/bind.h"
+#include "base/functional/callback_forward.h"
+#include "base/functional/callback_helpers.h"
+#include "brave/components/constants/pref_names.h"
+#include "components/content_settings/core/common/content_settings_pattern.h"
+#include "components/permissions/features.h"
+#include "components/permissions/permissions_client.h"
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/web_contents.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "third_party/blink/public/mojom/permissions/permission_status.mojom-shared.h"
+
+namespace permissions {
+
+namespace {
+
+bool IsGoogleAuthUrl(const GURL& gurl) {
+  // Check if pattern matches the URL
+  auto google_pattern = ContentSettingsPattern::FromString(kGoogleAuthPattern);
+  auto firebase_pattern = ContentSettingsPattern::FromString(kFirebasePattern);
```

Comment:
can we also cache these? Currently these are created for each main frame request.
please make `MaybeCreateThrottleFor` as fast as possible.

Reply:
I cached a bunch, let me know what you think